src/persist-txn/src/operator.rs

```diff
@@ -0,0 +1,422 @@
+        let DataSubscribe {
+            worker: _,
+            data,
+            txns,
+            capture: _,
+            output,
+            _tokens: _,
+        } = self;
+        f.debug_struct("DataSubscribe")
+            .field("data", data)
+            .field("txns", txns)
+            .field("output", output)
+            .finish_non_exhaustive()
+    }
+}
+
+impl DataSubscribe {
+    /// Creates a new [DataSubscribe].
+    pub fn new(
+        name: &str,
+        client: PersistClient,
+        txns_id: ShardId,
+        data_id: ShardId,
+        as_of: u64,
+    ) -> Self {
+        let mut worker = Worker::new(WorkerConfig::default(), Thread::new());
+        let (data, txns, capture, tokens) =
+            worker.dataflow::<u64, _, _>(|scope| {
+                let (data_stream, token0) = scope.scoped::<u64, _, _>("hybrid", |scope| {
+                    let client = client.clone();
+                    let (data_stream, token) = shard_source::<String, (), u64, i64, _, _, _, _>(
+                        scope,
+                        name,
+                        move || std::future::ready(client.clone()),
+                        data_id,
+                        Some(Antichain::from_elem(as_of)),
+                        Antichain::new(),
+                        false.then_some(|_, _: &_, _| unreachable!()),
+                        Arc::new(StringSchema),
+                        Arc::new(UnitSchema),
+                        |_| true,
+                    );
+                    (data_stream.leave(), token)
+                });
+                let (mut data, mut txns) = (ProbeHandle::new(), ProbeHandle::new());
+                let data_stream = data_stream.flat_map(|part| {
+                    part.map(|((k, v), t, d)| {
+                        let (k, ()) = (k.unwrap(), v.unwrap());
+                        (k, t, d)
+                    })
+                });
+                let data_stream = data_stream.probe_with(&mut data);
+                let (data_stream, token1) =
+                    txns_progress::<String, (), u64, i64, _, TxnsCodecDefault, _>(
+                        data_stream,
+                        name,
+                        std::future::ready(client.clone()),
+                        txns_id,
+                        data_id,
+                        as_of,
+                        Arc::new(StringSchema),
+                        Arc::new(UnitSchema),
+                    );
+                let data_stream = data_stream.probe_with(&mut txns);
+                let tokens: Rc<dyn Any> = Rc::new((token0, token1));
+                (data, txns, data_stream.capture(), tokens)
+            });
+        Self {
+            worker,
+            data,
+            txns,
+            capture,
+            output: Vec::new(),
+            _tokens: tokens,
+        }
+    }
+
+    /// Returns the exclusive progress of the dataflow.
+    pub fn progress(&self) -> u64 {
```

Comment:
Uncovered according to coverage report. Feel free to ignore this finding if considered too trivial to test.

Reply:
It's trivial but part of the intended public API, so worth having at least some coverage. Done!

Reply:
Thank you!